### PR TITLE
Send `Sec-Purpose: prefetch` header for prefetch requests

### PR DESCRIPTION
**Pull request by noamr**
Send `Sec-Purpose: prefetch` for prefetch requests

The purpose headers allows servers to discern between regular requests
and "resource hints" - requests that are only meant to populate
caches.

This replaces the existing `Purpose` and `x-moz: prefetch` headers.

Closes w3c/webappsec-fetch-metadata#84
Closes https://github.com/w3c/resource-hints/issues/74
Part of whatwg/html#8111

- [x] At least two implementers are interested (and none opposed):
   * Chromium
   * Apple: https://github.com/WebKit/standards-positions/issues/114
   * https://github.com/mozilla/standards-positions/issues/723

- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/35707
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: 
                     [1358419: `Sec-Purpose` header](https://bugs.chromium.org/p/chromium/issues/detail?id=1358419)
   * Firefox: 
                    [1788167: Align prefetch with spec](https://bugzilla.mozilla.org/show_bug.cgi?id=1788167)
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=49238
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/23553

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/fetch/1576.html" title="Last updated on Jan 16, 2023, 1:15 PM UTC (ac3e6d4)">Preview</a> | <a href="https://whatpr.org/fetch/1576/4500d3a...ac3e6d4.html" title="Last updated on Jan 16, 2023, 1:15 PM UTC (ac3e6d4)">Diff</a>

### CODE REVIEW

**Comment by mnot** [comment]
If this lands, please don't forget to [register the header](https://github.com/protocol-registries/http-fields#readme).

**Comment by annevk** [comment]
That's a good point. Let's define this header then in its own section in HTTP extensions?

Also, we should probably define it as a structured header given that its new and use https://fetch.spec.whatwg.org/#concept-header-list-set-structured-header to set it.

**Comment by noamr** [comment]
> That's a good point. Let's define this header then in its own section in HTTP extensions?
> 
> Also, we should probably define it as a structured header given that its new and use https://fetch.spec.whatwg.org/#concept-header-list-set-structured-header to set it.

Done, hope this is right :)

**Comment by noamr** [comment]
@annevk all the checkboxes for this and for https://github.com/whatwg/html/pull/8111 are ticked! Anything missing?

File: fetch.bs [hunk 1]
```diff
@@ -4088,6 +4090,26 @@ run these steps:
...
 
 
+<h3 id="sec-purpose-header">`<code>Sec-Purpose</code>` header</h3>
+
+<p>The `<dfn export http-header id=http-sec-purpose><code>Sec-Purpose</code></dfn>` HTTP request
+header specifies that the request serves one or more purposes other than requesting the resource for
+immediate use by the user.
+
+<p>The `<a http-header><code>Sec-Purpose</code></a>` header field is a <a>structured header</a>
+whose value must be a <a data-lt="structured header list">list</a>. Its ABNF is: [[!RFC8941]]
```
Comment (annevk) [inline fetch.bs:4100]: I think for now we can just state that it has to be a token? Also no need for ABNF.

File: fetch.bs [hunk 2]
```diff
@@ -4088,6 +4090,26 @@ run these steps:
...
+
+<pre><code class=lang-abnf>
+Sec-Purpose    = sf-list
+</code></pre>
+
+It may contain an item member which is the token "prefetch". If so, this indicates the request’s
```
Comment (annevk) [inline fetch.bs:4106]: Needs `<code>` as well.

File: fetch.bs [hunk 3]
```diff
@@ -4088,6 +4090,26 @@ run these steps:
...
+purpose is to download a resource it is anticipated will be fetched shortly.
+
+<p class=note>This can be used, for example, to let the server adjust the caching expiry for
+prefetches, to disallow the prefetch by rejecting the request, or treat it differently
+for the purpose of counting page visits.
+
```
Comment (annevk) [inline fetch.bs:4112]: You're eating a newline.

File: fetch.bs [hunk 4]
```diff
@@ -5313,6 +5335,10 @@ run these steps:
    <li><p><a abstract-op lt="append the Fetch metadata headers for a request">Append the Fetch metadata headers for <var>httpRequest</var></a>.
    [[!FETCH-METADATA]]
 
+   <li><p>If <var>httpRequest</var>'s <a for=request>initiator</a> is "<code>prefetch</code>", then
+   <a>set a structured field value</a> given (`<a http-header><code>Sec-Purpose</code></a>`,
+   "<code>prefetch</code>") in <var>httpRequest</var>'s <a for=request>header list</a>.
```
Comment (annevk) [inline fetch.bs:5340]: We're hitting a problem here with this API and ambiguity between strings and tokens.

We could omit the quotes, but that's not really a thing we normally define. Perhaps we just write it out as:

> token "`prefetch`"

with an xref for "token"?

@domenic thoughts?

Comment (domenic) [reply fetch.bs:5338]: I would do

> the token `prefetch`

I think it's confusing to have both "token" as the marker (indicating a token), and the quotation marks as the marker (indicating a string).

**Review by annevk — APPROVED** [review]
I pushed some more commits. This looks good now. (Once the HTML PR is approved I will merge this.)

**Comment by annevk** [comment]
Header registration: https://github.com/protocol-registries/http-fields/issues/37.

**Comment by jeremyroman** [comment]
The WebKit bug link looks incorrect; did you intend to link a different bug?

**Comment by annevk** [comment]
@jeremyroman thank you, I found a more suitable reference.

**Pull request merged**